File: mahaam-api-py/feat/plan/plan_service.py
from uuid import UUID
from typing import List, Protocol
from feat.plan.plan_model import Plan, PlanType
from feat.plan.plan_members_repo import PlanMembersRepo
from feat.user.suggested_emails_repo import SuggestedEmailsRepo
from feat.user.user_repo import UserRepo
from infra.exceptions import InputException, LogicException, UnauthorizedException
from infra.log import Log
from infra.validation import ProtocolEnforcer
from feat.plan.plan_model import PlanIn
from feat.plan.plan_repo import PlanRepo
from infra import db
from infra.req import req


class PlanService(Protocol):
    def select_one(self, plan_id: UUID) -> Plan: ...
    def select_many(self, type: str | None) -> list[Plan]: ...
    def create(self, plan: PlanIn) -> UUID: ...
    def update(self, plan: PlanIn) -> None: ...
    def delete(self, id: UUID) -> None: ...
    def share(self, id: UUID, email: str) -> None: ...
    def unshare(self, id: UUID, email: str) -> None: ...
    def leave(self, id: UUID) -> None: ...
    def update_type(self, id: UUID, type: str) -> None: ...
    def reorder(self, type: str, old_order: int, new_order: int) -> None: ...
    def validate_user_owns_the_plan(self, plan_id: UUID) -> None: ...
    def _validate_user_logged_in(self) -> None: ...


class DefaultPlanService(metaclass=ProtocolEnforcer, protocol=PlanService):
    def __init__(self, plan_repo: PlanRepo, plan_members_repo: PlanMembersRepo, user_repo: UserRepo, suggested_emails_repo: SuggestedEmailsRepo) -> None:
        self.plan_repo = plan_repo  
        self.plan_members_repo = plan_members_repo
        self.user_repo = user_repo
        self.suggested_emails_repo = suggested_emails_repo
        
              
    def select_one(self, plan_id: UUID) -> Plan:
        plan = self.plan_repo.select_one(plan_id)
        if plan and plan.is_shared:
            plan.members = self.plan_members_repo.get_users(plan_id)
        return plan

    def select_many(self, type: str | None) -> list[Plan]:
        # plans of the user shared or not
        plans = self.plan_repo.select_many(req.user_id, type)
        if req.is_logged_in:
            # plans of others shared with the user
            plan_members = self.plan_members_repo.get_other_plans(req.user_id)
            plans.extend(plan_members)
        return plans

    def create(self, plan: PlanIn) -> UUID:
        count = self.plan_repo.get_count(req.user_id, PlanType.MAIN)
        if count >= 100:
            raise LogicException("max_is_100", "Max is 100")

        return self.plan_repo.create(plan)

    def update(self, plan: PlanIn) -> None:
        self.validate_user_owns_the_plan(plan.id)
        self.plan_repo.update(plan)

    def delete(self, id: UUID) -> None:
        self.validate_user_owns_the_plan(id)
        with db.DB.transaction_scope() as conn:
            self.plan_repo.remove_from_order(req.user_id, id, conn)
            # This will delete all related records as it is a cascade delete
            self.plan_repo.delete(id, conn)

    def share(self, id: UUID, email: str) -> None:
        self._validate_user_logged_in()
        self.validate_user_owns_the_plan(id)
        user = self.user_repo.select_one_by_email(email)
        if user is None:
            raise InputException(f"email:{email} was not found")

        if user.id == req.user_id:
            raise LogicException(
                "not_allowed_to_share_with_creator", "Not allowed to share with creator")

        limit = 20
        plan = self.plan_repo.select_one(id)
        if plan and plan.is_shared:
            user_id_count = self.plan_members_repo.get_users_count(id)
            if user_id_count >= limit:
                raise LogicException("max_is_20", "Max is 20")
        else:
            plans_count = self.plan_members_repo.get_plans_count(req.user_id)
            if plans_count >= limit:
                raise LogicException("max_is_20", "Max is 20")

        self.plan_members_repo.create(id, user.id)
        self.suggested_emails_repo.create(req.user_id, email)
        usr = self.user_repo.select_one(req.user_id)
        self.suggested_emails_repo.create(user.id, usr.email)

    def unshare(self, id: UUID, email: str) -> None:
        self._validate_user_logged_in()
        self.validate_user_owns_the_plan(id)
        user = self.user_repo.select_one_by_email(email)
        if user is None:
            raise InputException(f"email:{email} was not found")

        self.plan_members_repo.delete(id, user.id)

    def leave(self, id: UUID) -> None:
        self._validate_user_logged_in()
        deleted_records = self.plan_members_repo.delete(id, req.user_id)
        if deleted_records == 1:
            Log.info(f"user {req.user_id} left plan {id}")
        else:
            raise InputException(f"userId={req.user_id} cannot leave planId={id}")

    def update_type(self, id: UUID, type: str) -> None:
        self.validate_user_owns_the_plan(id)
        count = self.plan_repo.get_count(req.user_id, type)
        if count >= 100:
            raise LogicException("max_is_100", "Max is 100")

        with db.DB.transaction_scope() as conn:
            self.plan_repo.remove_from_order(req.user_id, id, conn)
            self.plan_repo.update_type(req.user_id, id, type, conn)

    def reorder(self, type: str, old_order: int, new_order: int) -> None:
        count = self.plan_repo.get_count(req.user_id, type)
        if old_order > count or new_order > count:
            raise InputException("oldOrder and newOrder should be less than " + count)
        self.plan_repo.update_order(req.user_id, type, old_order, new_order)

    def validate_user_owns_the_plan(self, plan_id: UUID) -> None:
        plan = self.plan_repo.select_one(plan_id)
        if plan is None:
            raise InputException("planId not found")
        if plan.user.id != req.user_id:
            raise UnauthorizedException("User does not own this plan")

    def _validate_user_logged_in(self) -> None:
        user = self.user_repo.select_one(req.user_id)
        if user.email is None:
            raise LogicException("you_are_not_logged_in",
                                 "You are not logged in")

File: mahaam-api-java/src/main/java/mahaam/feat/plan/PlanService.java
package mahaam.feat.plan;

import java.util.List;
import java.util.UUID;

import jakarta.enterprise.context.ApplicationScoped;
import jakarta.inject.Inject;
import jakarta.transaction.Transactional;
import mahaam.feat.plan.PlanModel.Plan;
import mahaam.feat.plan.PlanModel.PlanIn;
import mahaam.feat.user.SuggestedEmailsRepo;
import mahaam.feat.user.UserModel.User;
import mahaam.feat.user.UserRepo;
import mahaam.infra.Exceptions.InputException;
import mahaam.infra.Exceptions.LogicException;
import mahaam.infra.Exceptions.UnauthorizedException;
import mahaam.infra.Log;
import mahaam.infra.Req;

public interface PlanService {
	Plan getOne(UUID planId);

	List<Plan> getMany(String type);

	UUID create(PlanIn plan);

	void update(PlanIn plan);

	void delete(UUID id);

	void share(UUID id, String email);

	void unshare(UUID id, String email);

	void leave(UUID id);

	void updateType(UUID id, String type);

	void reOrder(String type, int oldOrder, int newOrder);

	void validateUserOwnsThePlan(UUID planId);
}

@ApplicationScoped
class DefaultPlanService implements PlanService {

	@Inject
	PlanRepo planRepo;

	@Inject
	PlanMembersRepo planMembersRepo;

	@Inject
	UserRepo userRepo;

	@Inject
	SuggestedEmailsRepo suggestedEmailsRepo;

	@Override
	public Plan getOne(UUID planId) {
		Plan plan = planRepo.getOne(planId);
		if (plan != null && plan.isShared) {
			List<User> members = planMembersRepo.getUsers(planId);
			// Since Plan is now a class (mutable), we can modify fields directly
			plan.members = members;
		}
		return plan;
	}

	@Override
	public List<Plan> getMany(String type) {
		List<Plan> plans = planRepo.getMany(Req.getUserId(), type); // plans of the user shared
																	// or not
		if (Req.isLoggedIn()) {
			// plans of others shared with the user
			List<Plan> sharedPlans = planMembersRepo.getOtherPlans(Req.getUserId());
			plans.addAll(sharedPlans);
		}
		return plans;
	}

	@Override
	public UUID create(PlanIn plan) {
		var count = planRepo.getCount(Req.getUserId(), "Main");
		if (count >= PlanConstants.MAX_GROUPS_PER_USER) {
			throw new LogicException("max_is_100", "Max is 100");
		}

		return planRepo.create(plan);
	}

	@Override
	public void update(PlanIn plan) {
		validateUserOwnsThePlan(plan.id);
		planRepo.update(plan);
	}

	@Override
	@Transactional
	public void delete(UUID id) {
		validateUserOwnsThePlan(id);

		planRepo.removeFromOrder(Req.getUserId(), id);
		planRepo.delete(id); // This will delete all related records as it is a cascade delete
	}

	@Override
	public void share(UUID id, String email) {
		validateUserLoggedIn();
		validateUserOwnsThePlan(id);
		User user = userRepo.getOne(email);
		if (user == null) {
			throw new InputException("email_not_found: email:" + email + " was not found");
		}

		if (user.id.equals(Req.getUserId())) {
			throw new LogicException("not_allowed_to_share_with_creator", "Not allowed to share with creator");
		}

		Plan plan = planRepo.getOne(id);
		if (plan != null && plan.isShared) {
			var membersCount = planMembersRepo.getUsersCount(id);
			if (membersCount >= PlanConstants.MAX_SHARED_USERS) {
				throw new LogicException("max_is_20", "Max is 20");
			}
		} else {
			var plansCount = planMembersRepo.getPlansCount(Req.getUserId());
			if (plansCount >= PlanConstants.MAX_SHARED_GROUPS) {
				throw new LogicException("max_is_20", "Max is 20");
			}
		}

		planMembersRepo.create(id, user.id);
		suggestedEmailsRepo.create(Req.getUserId(), email);
		User usr = userRepo.getOne(Req.getUserId());
		suggestedEmailsRepo.create(user.id, usr.email);
	}

	@Override
	public void unshare(UUID id, String email) {
		validateUserLoggedIn();
		validateUserOwnsThePlan(id);
		User user = userRepo.getOne(email);
		if (user == null) {
			throw new InputException("email_not_found: email:" + email + " was not found");
		}

		planMembersRepo.delete(id, user.id);
	}

	@Override
	public void leave(UUID id) {
		validateUserLoggedIn();
		int deletedRecords = planMembersRepo.delete(id, Req.getUserId());
		if (deletedRecords == 1) {
			Log.info("user " + Req.getUserId() + " left plan " + id);
		} else {
			throw new InputException("userId=" + Req.getUserId() + " cannot leave planId=" + id);
		}
	}

	@Override
	@Transactional
	public void updateType(UUID id, String type) {
		validateUserOwnsThePlan(id);
		var count = planRepo.getCount(Req.getUserId(), type);
		if (count >= 100) {
			throw new LogicException("max_is_100", "Max is 100");
		}

		planRepo.removeFromOrder(Req.getUserId(), id);
		planRepo.updateType(Req.getUserId(), id, type);
	}

	@Override
	public void reOrder(String type, int oldOrder, int newOrder) {
		var count = planRepo.getCount(Req.getUserId(), type);
		if (oldOrder > count || newOrder > count) {
			throw new InputException("oldOrder and newOrder should be less than " + count);
		}
		planRepo.updateOrder(Req.getUserId(), type, oldOrder, newOrder);
	}

	@Override
	public void validateUserOwnsThePlan(UUID planId) {
		Plan plan = planRepo.getOne(planId);
		if (plan == null) {
			throw new InputException("planId not found");
		}
		if (!plan.user.id.equals(Req.getUserId())) {
			throw new UnauthorizedException("User does not own this plan");
		}
	}

	private void validateUserLoggedIn() {
		if (!Req.isLoggedIn()) {
			throw new LogicException("you_are_not_logged_in", "You are not logged In");
		}
	}
}

class PlanConstants {
	public static final int MAX_GROUPS_PER_USER = 100;
	public static final int MAX_SHARED_USERS = 20;
	public static final int MAX_SHARED_GROUPS = 20;
}